contracts/src/SafeInternationalHarbour.sol

```diff
@@ -0,0 +1,406 @@
+        address signerAddress,
+        address safeAddress,
+        uint256 chainId,
+        uint256 nonce
+    ) external view returns (uint256 count) {
+        count = _sigData[signerAddress][safeAddress][chainId][nonce].length;
+    }
+
+    /**
+     * @notice Computes the unique EIP-712 digest for a SafeTx using the provided parameters and domain.
+     * @param safeAddress Address of the target Safe Smart Account.
+     * @param chainId Chain ID included in the domain separator.
+     * @param nonce Safe transaction nonce.
+     * @param to Target address the Safe will call.
+     * @param value ETH value to be sent with the call.
+     * @param data Call data executed by the Safe.
+     * @param operation Operation type: 0 = CALL, 1 = DELEGATECALL.
+     * @param safeTxGas Gas limit for the Safe's internal execution.
+     * @param baseGas Base gas overhead for reimbursement.
+     * @param gasPrice Gas price used for reimbursement calculation.
+     * @param gasToken Token address for refunds (0x0 for ETH).
+     * @param refundReceiver Address to receive gas refunds.
+     * @return safeTxHash Keccak256 digest of the EIP-712 encoded SafeTx.
+     */
+    function _computeSafeTxHash(
+        address safeAddress,
+        uint256 chainId,
+        uint256 nonce,
+        address to,
+        uint256 value,
+        bytes calldata data,
+        uint8 operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address refundReceiver
+    ) private pure returns (bytes32 safeTxHash) {
+        bytes32 domainSeparator = keccak256(
+            abi.encode(_DOMAIN_TYPEHASH, chainId, safeAddress)
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(
+                _SAFE_TX_TYPEHASH,
+                to,
+                value,
+                keccak256(data),
+                operation,
+                safeTxGas,
+                baseGas,
+                gasPrice,
+                gasToken,
+                refundReceiver,
+                nonce
+            )
+        );
+        safeTxHash = keccak256(abi.encodePacked("\x19\x01", domainSeparator, structHash));
+    }
+
+    /**
+     * @notice Splits a 65-byte ECDSA signature into its components and recovers the signer address.
+     * @param digest The message or data hash to verify (EIP-712 digest or eth_sign prefixed).
+     * @param sig Concatenated 65-byte ECDSA signature (r || s || v).
+     * @return signer The address that produced the signature (EOA).
+     * @return r First 32 bytes of the ECDSA signature.
+     * @return s Second 32 bytes of the ECDSA signature.
+     * @dev Supports both EIP-712 and eth_sign flows by detecting v > 30 and applying the Ethereum Signed Message prefix.
+     */
+    function _recoverSignerAndRS(bytes32 digest, bytes calldata sig)
+        private
+        pure
+        returns (address signer, bytes32 r, bytes32 s)
```

Comment:
The `v` is important for signature recovery and can be packed into the `s` value by normalizing it to half of the curve order.

This way, you only use 2 storage slots, and prevent signature malleability issues.

Reply:
but there are only four values to choose from, right? So the client can just brute force them, that was my plan

but since there is a [standardised](https://eips.ethereum.org/EIPS/eip-2098) compact form, I can also use it (but I'm not sure how well it is supported in client libraries

Reply:
To ensure I understand everything, how does it prevent signature malleability issues?